OIP-4: eWallet release and versioning

sirn opened the pull request:
Request for some initial comment. Still lack details and other stuff. If you need me to clarify on other things or want me to add anything please comment.

I ultimately decided to drop the Long-Term Release for now as I don't think we have enough manpower to do it. Some ideas of the original LTS proposal is now merged into a stable release which should allow us to have something stable for enterprise.

I'm still very unsure about the branching model. I feel it's a bit too complex. On the other hand I feel having stable branch back-porting changes from develop allow us to move fast with develop, while give us a stable base to work on for a release without the need to hold on PRs merge on develop when we need to release.

Anyway, open to comments. Sorry for really late on this.

jarindr commented:
I didn't read into detail yet, but do you consider using `release` feature from github to mark it as stable version ?

sirn commented:
GitHub releases are associated with a tag so yes :-)

whoisjeremylam commented:
Would we be able to expand on what `dark ship` means and what would qualify to be merged in that manner?

I quickly searched Google and wasn't able to find anything...

kasima commented:
Sorry, I guess "dark ship" doesn't seem like the common term. Looks like people use ["dark launch"](https://www.google.com/search?q=dark+launch+software&oq=dark+launch+software&aqs=chrome..69i57.3560j0j7&sourceid=chrome&ie=UTF-8).

Here's a description from one of the results:

> Dark launching is the process of releasing production-ready features to a subset of your users prior to a full release. This enables you to decouple deployment from release...

In this particular case, we would be releasing stuff to 0 users. But it would allow us to merge features quickly (as long as we were responsible with how we implement it) but not expose them if we didn't want them to go out in a release.

sirn commented:
My concern about dark ship is when it touches too much code and cause regression. Though, maintaining a separate feature branch for big changes and never merge until it's stable might be a better idea indeed.

kasima commented:
> My concern about dark ship is when it touches too much code and cause regression.

Absolutely. Anything merged needs to basically be production-ready. It's there as an option if someone feels like that's better for what they're developing.

kasima commented:
I was also thinking that maybe we don't even need to keep `supported-branch-1` and `supported-branch-2` around. Just work from the tags, when needed.

T-Dnzt commented:
@kasima @sirn I'm a bit worried about removing the intermediate `stable` branch, based on bugs that we've seen happening only after `develop` had been deployed to staging and had to be fixed there (even with everything working perfectly in local). I feel we cannot be sure everything we're merging to develop is production-ready basically.

With `develop` + `stable`, it feels safer and we're sure no one will be using `develop` directly. Everything in `stable` should, however, be production ready (and has been tested and deployed in a live environment (staging)). I understand "dark shipping" is that it's not always applicable without spending more time making it possible - not sure what is more time consuming between having a more complex branching system and having "dark shipping" capabilities.

About not keeping branches for each stable release and only using tags, I don't really mind as long as it's as easy for us to cherry pick and add fixes. I guess it removes the step of dealing with the branch since we would branch out from the tag and update it directly. Why not :)

kasima commented:
> Everything in stable should, however, be production ready (and has been tested and deployed in a live environment (staging))

This could still be true using tags and short lived release branches as well.

The main worry I have is cherry-picking from `develop` into `staging` since it can be time consuming and can introduce regressions on its own. You're right though – it comes down to trade-offs between:

1. making a PR safe enough to dark ship
1. having long running feature branches
1. managing cherry-picking / conflicts for every release

I comes down to what work you'd prefer to do. Personally, after working in `svn` for years and managing merges and conflicts for release, I never want to do that again – even in `git`. But that is personal preference. Also, I know that I just outed myself as 👴🏽

sirn commented:
If we're going with a long-running feature branch, I think having a per-feature deploy (e.g. `feature-branch-a` automatically deploys to `feature-branch-a.staging.omisego.io` or something) with data copy & migrate from master staging would be great too. It also has a nice side-effect of encouraging everyone to rebase to master as often as possible.

kasima commented:
> having a per-feature deploy (e.g. feature-branch-a automatically deploys to feature-branch-a.staging.omisego.io or something) with data copy & migrate from master staging would be great too

😍 

Instead of data copy, maybe a fixed set of seed/sample data?

sirn commented:
> Instead of data copy, maybe a fixed set of seed/sample data?

I thought about that, but given all the weird data we have in staging, I think we better use something that resembles an actual (ab)use so we could cover all edge cases (e.g. during migrations).

T-Dnzt commented:
Love the idea of auto-deploying feature branches, if we do that, I don't mind getting rid of the `stable`, so let's make it happen!  

For automatic deploys, having the option to either start fresh with seeds (regular, sample or e2e tests) or copy from another DB (staging, demo, etc) would be awesome!

@sirn can you update the docs with all the feedback so we can approve and proceed :) Thanks!

sirn commented:
I'll update the doc. In the mean time, I have to take a look at how to do feature branch deploy first 🤦‍♂️

sirn commented:
Updated per comments.

T-Dnzt commented on xxxx-release-and-versioning.md:
```diff
@@ -27,45 +37,51 @@ Only a the latest `MAJOR.MINOR.PATCH` version of the last 2 `MAJOR.MINOR` versio
-Stable branch shall be separated from the mainline development branch with changes selectively backported when a feature is considered stable enough. Backporting shall be done via a **Pull Request** against the stable branch and shall be done on a per-feature basis.
-
-Once a stable release will happen, a stable release shall have its own branch in a form of `stable-MAJOR.MINOR` and tag with `vMAJOR.MINOR.PATCH` as appropriate.
+The branching model will utilize a `master` branch, a stable version branch and a feature branch. To demostrate:
```
`to demonstrate`

T-Dnzt replied:
To demonstrate:

T-Dnzt commented on xxxx-release-and-versioning.md:
```diff
@@ -27,45 +37,51 @@ Only a the latest `MAJOR.MINOR.PATCH` version of the last 2 `MAJOR.MINOR` versio
-In this chart, changes are selectively backported from `develop` to `stable` in a per-feature basis (`{a1,a2}`, `{b1,b2}`, ...) This process can be done by cherry-picking changes and fixing conflicts as necessary.
+#### Feature Branches
+
+Feature branches, such as `feature2` and `feature1`, are development branches that implements new feature or change existing functionality. These branches may be long-running and should be merged to `master` only when the feature is considered stable enough for public consumption. Feature branches should regularly sync up with master to avoid too much code diversion. Feature branches may have its ...
```
Feature branches may each have their own public testing environment with data copied from staging on each deploy.

T-Dnzt commented on xxxx-release-and-versioning.md:
```diff
@@ -6,15 +6,25 @@ This document specifies a procedure for releasing and versioning to allow for st
+
+A stable release will be provided as a Docker image based on Alpine Linux. The Docker image will be tagged with the release version (e.g. `v1.0.3`) and will use a `:latest` tag to refer to the latest stable release.
+
+Only the latest `MAJOR.MINOR.PATCH` version of the last 2 `MAJOR.MINOR` versions will be supported. However, we may provide security fixes to non-supported versions as the team see fit. For demostration purpose, given that we released the following versions at a specific date, our support condition shall be:
```
For demonstration purposes, given that we released the following versions at specific dates, our support condition shall be:

T-Dnzt commented on xxxx-release-and-versioning.md:
```diff
@@ -27,45 +37,51 @@ Only a the latest `MAJOR.MINOR.PATCH` version of the last 2 `MAJOR.MINOR` versio
+
+#### `master` branch
+
+The `master` branch is the branch that development release will be based on. While we required that feature branch must be stable enough before it could be merged into `master`, the `master` branch may still not stable enough to use in production environment. The `master` branch is where a stable branch will be branched from when it is considered stable enough. Before branching into stable, `mast ...
```
While we require that feature branches must be stable enough before it getting merged into `master`, the `master` branch may still not be stable enough to use in a production environment.

kasima replied:
My sense is that people working in production would understand that `master` is latest and greatest and versioned releases are for production. At least this is what I see in the ruby/rails/js/node worlds.

T-Dnzt commented on xxxx-release-and-versioning.md:
```diff
@@ -27,45 +37,51 @@ Only a the latest `MAJOR.MINOR.PATCH` version of the last 2 `MAJOR.MINOR` versio
+
+Feature branches, such as `feature2` and `feature1`, are development branches that implements new feature or change existing functionality. These branches may be long-running and should be merged to `master` only when the feature is considered stable enough for public consumption. Feature branches should regularly sync up with master to avoid too much code diversion. Feature branches may have its ...
+
+#### `master` branch
```
I feel having `master` has development branch is a bit weird, but it's probably my personal experience talking. Using `develop` would probably make it clearer for people that they shouldn't use that branch directly 🤔

T-Dnzt replied:
At the same time, having a repo without `master` would probably be weird 🤷‍♂️

unnawut commented on xxxx-release-and-versioning.md:
```diff
@@ -0,0 +1,87 @@
+| End to end testing  | **Yes**  |
+| Migration testing   | **Yes**  |
+
+A stable release will be provided as a Docker image based on Alpine Linux. The Docker image will be tagged with the release version (e.g. `v1.0.3`) and will use a `:latest` tag to refer to the latest stable release.
```
No binary release for stable?

sirn replied:
I thought about it, then decided that stable should be Docker-only for now (as well as maybe nightly). We simply don't have enough manpower to build and test more than single platform right now, and releasing an official binary for N platforms is a little bit complicated. We might do in the future though.

unnawut commented on xxxx-release-and-versioning.md:
```diff
@@ -0,0 +1,87 @@
+| 1.0.1   | 2017-08-01   | Not supported, `1.0` discontinued          |
+| 1.0.0   | 2016-08-01   | Not supported, `1.0` discontinued          |
+
+In the figure above, version `1.0.2` is released as a security fix outside of the support period. This release is considered a hotfix release. However, even though version `1.0.2` is released after the supported version `1.1.1`, the version `1.0.2` won't be guarantee to receive any further updates. Provider is encouraged to upgrade to latest supported version.
```
I associate "hotfix" with fixes made to production environment (and somewhat quick and dirty), probably because of my past background. Should we clarify the definition a bit that:

> A hotfix is a patch that that addresses one or more specific issues on a version that is no longer supported.

?

sirn replied:
@T-Dnzt what term should we use? My thought is that "hotfix" means something that goes directly to release regardless whether they're still supported.

T-Dnzt replied:
I have the same understanding as you @sirn, I think `hotfix` is alright. Not limited to non-supported releases, maybe adding one more line to the table with a `hotfix` pushed on a supported release will clarify things.

unnawut commented on xxxx-release-and-versioning.md:
```diff
@@ -0,0 +1,87 @@
+
+#### Stable version branches
+
+Stable version branches such as `stable-1.0` and `stable-1.1`, are branches that correspond to the `MAJOR.MINOR` release versions. The branch is branched from `master` when the `master` branch is considered enable enough. Stable version branches are solely for maintenance purpose and will not accept any new features except for bug or security fixes.
```
Suggest naming version branches as  `v1.0`, `v1.1`, etc. instead, Similar to Elixir and quite a few other open source projects.

unnawut commented on xxxx-release-and-versioning.md:
```diff
@@ -0,0 +1,87 @@
+
+#### `master` branch
+
+The `master` branch is the branch that development release will be based on. While we required that feature branch must be stable enough before it could be merged into `master`, the `master` branch may still not stable enough to use in production environment. The `master` branch is where a stable branch will be branched from when it is considered stable enough. Before branching into stable, `mast ...
```
What do you think of doing release candidates? 

E.g. while aiming for a `1.1.0` release, also do `1.1.0-rc1`, `1.1.0-rc2`, ... Until no new bugs emerge we release `1.1.0`.

unnawut commented on xxxx-release-and-versioning.md:
```diff
@@ -0,0 +1,87 @@
+
+#### Feature Branches
+
+Feature branches, such as `feature2` and `feature1`, are development branches that implements new feature or change existing functionality. These branches may be long-running and should be merged to `master` only when the feature is considered stable enough for public consumption. Feature branches should regularly sync up with master to avoid too much code diversion. Feature branches may have its ...
```
@sirn @T-Dnzt Should we also mandate that feature branches are prefixed with `feature-`?

Also do we prefix our project-management or GitHub issue number in the branch?

sirn replied:
I and @T-Dnzt actually discussed about this tl;dr it's weird calling code cleanup a feature, and if we have multiple prefixes everyone would stop bothering about it at one point and just use feature/ for everything, which essentially the same as having no prefix.

unnawut commented:
By the way...

> While we required that feature branch must be stable enough before it could be merged into `master`, the `master` branch may still not stable enough to use in production environment. The `master` branch is where a stable branch will be branched from when it is considered stable enough.

Yes Yes Yes Yes Yes! This has been my biggest pain point so far regarding a stable branch during beta phase.

sirn commented:
I added RC and updated wording as suggested by @T-Dnzt. I'm not quite sure whether we'll consider the RC image as a release candidate (e.g. simply rename the `v1.0.0-rc.1` to `v1.0.0` on release) or we'll consider the _codebase_ as a release candidate (e.g. `v1.0.0-rc1` and `v1.0.0` has different checksums).

I'm incline on the latter, as rebuilding image may result in a bug [due to phase of the moon](https://en.wikipedia.org/wiki/Heisenbug), an actual bitrot, or more realistically a package version changes in the base image (even though we will build & freeze everything).

sirn commented:
I'm moving this to last call and will finalize on Wednesday if nobody has anything more to add.

sirn commented:
Merging!
